### ISSUE TITLE
Deleting a feature in vw_qgep_reach does not delete relative reach points

### ISSUE THREAD

**Issue by mdouchin** [QGEP/QGEP#333]
I have tried it in QGEP demo projet with QGEP demo data with theses steps (in order to merge without erging)
---[Oa]----R1-->--[Ob]----R2-->--[Oc]---

* Select vw_qgep_wastewater_structure in QGIS legend
* Toggle editing on, select [Ob] and delete [Ob] then toggle editing Off to save which leads to
---[Oa]----R1--->----|------R2-->--[Oc]---

* Open od_reach_point layer in QGIS
* Select reach R2 and delete it, then save via toggling editing off
-> the 2 features od_reach_point situated in the location of former [Ob] are not deleted.

There is a rule AFTER DELETE in vw_qgep_reach, which does not seem to be triggered. I think we should use TRIGGERS instead of rules. See 

```sql
CREATE OR REPLACE RULE vw_qgep_reach_on_delete AS
    ON DELETE TO qgep.vw_qgep_reach DO INSTEAD ( DELETE FROM qgep.od_reach
  WHERE od_reach.obj_id::text = old.obj_id::text;
 DELETE FROM qgep.od_wastewater_networkelement
  WHERE od_wastewater_networkelement.obj_id::text = old.obj_id::text;
 DELETE FROM qgep.od_reach_point
  WHERE od_reach_point.obj_id::text = old.rp_from_obj_id::text;
 DELETE FROM qgep.od_reach_point
  WHERE od_reach_point.obj_id::text = old.rp_to_obj_id::text;
);
```

**Comment by m-kuhn** [comment]
Hmm... confirmed here with plain sql on `vw_qgep_reach`.

Do you have an idea why it's not triggered? No problem to switch this particular rule to a trigger, reach and reach point is really a special case because of the 1:1 relation. At the same time I'l like to understand what's going wrong because *instead of rules* are used quite heavily throughout the db and if there's something about them which needs special treatment that would be important to know.

**Comment by 3nids** [comment]
you don't get any SQL error?
there is an AFTER DELETE trigger on od_reach, I don't see any issue here, but could it stop the rule?
@m-kuhn  can you raise notice to check what's going on?

**Comment by m-kuhn** [comment]
Nope, no error. The reach points just do not disappear.

**Comment by 3nids** [comment]
Can you try to add a RAISE NOTICE '%' , old.rp_from_obj_id; before deleting the reach point to see what happens (i.e. sql not reached or missing obj_id)?

**Comment by m-kuhn** [comment]
Doesn't work (in a rule?)

```
ERROR:  syntax error at or near "RAISE"
LINE 423:   RAISE NOTICE  '%' , old.rp_from_obj_id;
            ^
```

**Comment by 3nids** [comment]
indeed
https://www.postgresql.org/docs/9.2/static/sql-createrule.html

will look this afternoon

Le jeu. 15 juin 2017 à 11:52, Matthias Kuhn <notifications@github.com> a
écrit :

**Comment by 3nids** [comment]
https://stackoverflow.com/questions/5534927/how-to-write-a-delete-rule-on-a-view

I guess this is it. The old is deleted, hence you don't have access to reach points obj ids anymore.

Go for trigger, shall we put in place the [inherited view generator](https://github.com/opengisch/metaproject/tree/master/postgresql/pg_inheritance_view)?

**Comment by m-kuhn** [comment]
Trigger sounds good to me. So does the generator 👍

**Comment by mdouchin** [comment]
Good to me too. Thanks for the feedback

**Comment by urskaufmann** [comment]
Ist not only the reachpoints. if I delete a reach, then all other elements that are normally connected to the reach (wastewater networkelement, channel, wastewaterstructure) are also not deleted. 
With the wastewaterstructure theme its ok.

**Comment by m-kuhn** [comment]
The networkelement should be deleted by a trigger as well.

wastewater structure / channel: should we check if it was the last reach and if yes trash it as well?

**Comment by mdouchin** [comment]
Ok for trashing them, but we need to be sure the user can rollback, and at present the undo/redo is not possible.

**Comment by m-kuhn** [comment]
That's a fundamental problem, not isolated to ws/channel. It's fine for me to make this a pre-requisite if funds are available to fix this for real.

**Comment by urskaufmann** [comment]
Check if it was the last reach connected to the channel and if yes trash the channel/wastewaterstructure as well: is ok for me.
Deleting the wastewater networkelement and the reachpoints with every reach is necessary.

**Comment by mdouchin** [comment]
What about using more bullet proof FOREIGN KEY CONSTRAINT with ON DELETE CASCADE ?
The view trigger should only delete the main object, for example od_reach for the view vw_qgep_reach.
Then, the FOREIGN KEY rules will safely cascade the deletion.

I really think complex triggers (or rules) should be avoided if possible. The behaviour is not easy to understand for the end user (or even for more advanced users who should look at any view or table def to understand the behaviour)

See quick & dirty PNG at the bottom representing vw_qgep_reach. IMHO, we should add FK constraints for:

## vw_qgep_reach

od_reach:
* od_reach.fk_reach_point_from
* od_reach.fk_reach_point_to
* od_reach.fk_pipe_profile
* od_reach.obj_id (to cascade to od_wastewater_networkelement.obj_id)

od_reach_point:
* od_reach_point.fk_wastewater_networkelement

od_wastewater_networkelement:
* od_wastewater_networkelement.fk_wastewater_structure

od_wastewater_structure
* od_wastewater_structure.obj_id (to cascade to od_channel)

![vw_qgep_reach](https://user-images.githubusercontent.com/3492210/29415186-7544d3c6-8362-11e7-8e3a-8e4fac2acffe.png)

## vw_qgep_wastewaterstructure

todo: list needed FK constraint

![vw_qgep_wastewater_structure](https://user-images.githubusercontent.com/3492210/29415445-61e6cf04-8363-11e7-88eb-76e3f72bfc32.png)

**Comment by m-kuhn** [comment]
See also https://github.com/QGEP/QGEP/issues/143

**Comment by signedav** [comment]
As far I understand this issue, this is concerning the same like #436 and is solved by #87 like @sjib mentioned. So it can be closed.

Regarding the proposing sollution with FK ON DELETE CASCADE by @mdouchin 
That's for sure a proper way to make it the other way around. When we define on the `reach` the FK to the `reach_point` ON DELETE CASCADE the referencing `reach` is deleted on delete of the referenced `reach_point`. That's implemented meanwhile as well.
The rule that has been not working before has to exist as well to make the opposite: Delete the referenced `reach_point` when deleting the referencing `reach`. This works now.

So deleting both ways is provided.

**Comment by ponceta** [comment]
Thanks for the fix!
